Seneca-CDOT/starchart
# Add setup.sh script to create initial records in primary dns

Fixes #2.

This isn't complete yet, but I made more progress, and now it can create all the necessary records.  The replication to the secondary DNS is sort-of working, but I get an error.

To test this, do the following:

First, running the three containers together:

```sh
$ docker-compose up
[+] Running 4/4
 ⠿ Network starchart_private_net  Created                                                                                        0.0s
 ⠿ Container secondary            Created                                                                                        0.0s
 ⠿ Container web                  Created                                                                                        0.0s
 ⠿ Container primary              Created                                                                                        0.0s
Attaching to primary, secondary, web
web        | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
web        | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
web        | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
web        | 10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
web        | 10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
web        | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
web        | /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
web        | /docker-entrypoint.sh: Configuration complete; ready for start up
web        | 2022/11/27 02:36:43 [notice] 1#1: using the "epoll" event method
web        | 2022/11/27 02:36:43 [notice] 1#1: nginx/1.20.2
web        | 2022/11/27 02:36:43 [notice] 1#1: built by gcc 10.3.1 20210424 (Alpine 10.3.1_git20210424)
web        | 2022/11/27 02:36:43 [notice] 1#1: OS: Linux 5.15.49-linuxkit
web        | 2022/11/27 02:36:43 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker processes
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 32
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 33
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 34
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 35
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 36
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 37
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 38
web        | 2022/11/27 02:36:43 [notice] 1#1: start worker process 39
primary    | Nov 27 02:36:43 Loading '/usr/local/lib/pdns/libgsqlite3backend.so'
primary    | Nov 27 02:36:43 This is a standalone pdns
primary    | Nov 27 02:36:43 Listening on controlsocket in '/var/run/pdns/pdns.controlsocket'
primary    | Nov 27 02:36:43 UDP server bound to 10.5.0.20:53
primary    | Nov 27 02:36:43 TCP server bound to 10.5.0.20:53
primary    | Nov 27 02:36:43 PowerDNS Authoritative Server 4.7.2 (C) 2001-2022 PowerDNS.COM BV
primary    | Nov 27 02:36:43 Using 64-bits mode. Built using gcc 10.2.1 20210110 on Nov  2 2022 08:26:10 by root@52b1d43e0211.
primary    | Nov 27 02:36:43 PowerDNS comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it according to the terms of the GPL version 2.
primary    | Nov 27 02:36:43 [webserver] Listening for HTTP requests on 10.5.0.20:8081
secondary  | Nov 27 02:36:43 Loading '/usr/local/lib/pdns/libgsqlite3backend.so'
secondary  | Nov 27 02:36:43 This is a standalone pdns
secondary  | Nov 27 02:36:43 Listening on controlsocket in '/var/run/pdns/pdns.controlsocket'
secondary  | Nov 27 02:36:43 UDP server bound to 10.5.0.80:53
secondary  | Nov 27 02:36:43 TCP server bound to 10.5.0.80:53
secondary  | Nov 27 02:36:43 PowerDNS Authoritative Server 4.7.2 (C) 2001-2022 PowerDNS.COM BV
secondary  | Nov 27 02:36:43 Using 64-bits mode. Built using gcc 10.2.1 20210110 on Nov  2 2022 08:26:10 by root@52b1d43e0211.
secondary  | Nov 27 02:36:43 PowerDNS comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it according to the terms of the GPL version 2.
primary    | Nov 27 02:36:43 Polled security status of version 4.7.2 at startup, no known issues reported: OK
secondary  | Nov 27 02:36:43 [webserver] Listening for HTTP requests on 10.5.0.80:8081
secondary  | Nov 27 02:36:43 Polled security status of version 4.7.2 at startup, no known issues reported: OK
primary    | Nov 27 02:36:43 Creating backend connection for TCP
primary    | Nov 27 02:36:43 Primary/secondary communicator launching
primary    | Nov 27 02:36:43 About to create 3 backend threads for UDP
secondary  | Nov 27 02:36:43 Primary/secondary communicator launching
secondary  | Nov 27 02:36:43 Creating backend connection for TCP
secondary  | Nov 27 02:36:43 About to create 3 backend threads for UDP
primary    | Nov 27 02:36:44 Done launching threads, ready to distribute questions
secondary  | Nov 27 02:36:44 Done launching threads, ready to distribute questions
```

Next, in a second shell, run the `setup.sh` script on the `primary` DNS conatiner:

```sh
$ docker exec -it primary /home/pdns/setup.sh
Creating empty zone 'starchart.com'
New rrset:
starchart.com. 3600 IN NS ns2.starchart.com
New rrset:
starchart.com. 3600 IN NS ns2.starchart.com
starchart.com. 3600 IN NS ns1.starchart.com
New rrset:
ns1.starchart.com. 3600 IN A 10.5.0.20
New rrset:
ns2.starchart.com. 3600 IN A 10.5.0.80
New rrset:
www.starchart.com. 3600 IN A 10.5.0.100
Current records for starchart.com IN SOA will be replaced
New rrset:
starchart.com. 3600 IN SOA ns1.starchart.com mail.domain.com 1 10800 3600 604800 3600
Securing zone with default key size
Adding CSK (257) with algorithm ecdsa256
Zone starchart.com secured
Adding NSEC ordering information for zone 'starchart.com', 4 updates
SOA serial for zone starchart.com set to 2
Added to queue
Checked 6 records of 'starchart.com', 0 errors, 0 warnings.
Checked 1 zones, 0 had errors.
$ORIGIN .
ns1.starchart.com	3600	IN	A	10.5.0.20
ns2.starchart.com	3600	IN	A	10.5.0.80
starchart.com	3600	IN	NS	ns2.starchart.com.
starchart.com	3600	IN	NS	ns1.starchart.com.
starchart.com	3600	IN	SOA	ns1.starchart.com mail.domain.com 2 10800 3600 604800 3600
www.starchart.com	3600	IN	A	10.5.0.100
```

Notice the logs in the primary and secondary DNS containers as this happens:

```
primary    | Nov 27 02:37:58 Notification request for zone 'starchart.com' received from operator
secondary  | Nov 27 02:37:59 Received NOTIFY for starchart.com from 10.5.0.20 for which we are not authoritative, trying supermaster
secondary  | Nov 27 02:38:00 Error resolving SOA or NS for starchart.com at: 10.5.0.20: Query to '10.5.0.20' for SOA of 'starchart.com' produced no answers
```

So it's seeing the primary is notifying the secondary, and it tries to get the details, but `10.5.0.20` doesn't respond to the query for some reason.

I suspect it's something to do with PowerDNS not being the DNS running on that instance (i.e., my Docker host is likely providing the DNS instead).  I'm not sure how to override.
